python-backend/math_utils.py

```diff
@@ -11,41 +12,98 @@
     """
-    error calculation: log(|expression - L|)
-    :param expression: mathematical expression provided by user to evaluate
-    :param symbol: the variable to substitute in the expression
+    delta as defined by the expression -1 * (log(|Pn/Qn - L|) / log(Qn)) - 1
     :param limit: the limit of the expression toward positive infinity
     :param val: the x value and the value to substitute in for the symbol in the expression
-    :return: the error, i.e. y coordinate, at the val provided
+    :param denom_val: just the denominator of the compute value val
+    :return: the delta or y coordinate at the val provided
     """
-    return sympy.log(abs(expression.subs(symbol, val) - limit), 10).evalf(PRECISION)
+    return mpmath.mpf(-1) * mpmath.log10(abs(val - limit)) / mpmath.log10(denom_val) - mpmath.mpf(1)
 
 
-def log(expression: sympy.core, symbol: Symbol, val: int) -> sympy.core.numbers:
+def generalized_computed_values(a: sympy.core, b: sympy.core, symbol: Symbol, iterations: int = 500) -> (
+list[mpmath.mpf], list[mpmath.mpf]):
     """
-    log base 10 of expression at val
-    :param expression: mathematical expression provided by user to evaluate
-    :param symbol: the variable to substitute in the expression
-    :param val: the x value and the value to substitute in for the symbol in the expression
-    :return: the log base 10, i.e. y coordinate, at the val provided
+    Compute values at each step iteratively for a polynomial continued fraction
+    Parameters
+    ----------
+    :param iterations: computation depth
+    :param b: partial numerator provided by user
+    :param a: partial denominator provided by user
+    :param symbol: the variable used by the user in numerator and denominator
+
+    Returns
+    -------
+    list of values computed at each step from n = 1 to the number of iterations provided
     """
-    return sympy.log(expression.subs(symbol, val), 10).evalf(PRECISION)
+    # for complex a and b (note that the article reverses a and b)
+    # see https://en.wikipedia.org/wiki/Generalized_continued_fraction#
+    a_n = [mpmath.mpf(a.subs(symbol, n).evalf(PRECISION)) for n in range(0, iterations)]
+    b_n = [mpmath.mpf(b.subs(symbol, n).evalf(PRECISION)) for n in range(0, iterations)]
+    for i in range(0, len(a_n)):
+        logger.debug(f"a value at {i}: {a_n[i]}, b value at {i}: {b_n[i]}")
+    # these arrays start at n = -1, so we prepad the convergent list with values we will slice off
+    numerators = [mpmath.mpf(1), a_n[0]]
+    denominators = [0, mpmath.mpf(1)]
+    convergents = [None, numerators[1] / denominators[1]]
+    # note n >= 1 per the article above
+    for i in range(1, len(b_n)):
+        numerators.append(a_n[i] * numerators[i] + b_n[i] * numerators[i - 1])
+        denominators.append(a_n[i] * denominators[i] + b_n[i] * denominators[i - 1])
 
+        if denominators[i] != 0:
+            convergents.append(numerators[-1] / denominators[-1])
+            if i < 20:
```

Comment:
What is significant about 20? Consider making this a variable whose name indicates the significance.

Reply:
It's arbitrary for debugging purposes. I didn't want to print 5000 values. The first few are all you need to debug. I will add a comment, however, because that is not clear and given all the complex math it would make that clear.